### PR TITLE
feat(scripts): fix bug about `sed -i`

### PR DESCRIPTION
**Pull request by mask-pp**
## Issue Addressed

Which issue # does this PR address?
appear error when run `./setup_time.sh genesis.json`: 

- `grep: invalid option -- P`
- `sed: 1: "genesis.json": extra characters at the end of g command`
- typo: `taif` in `scripts/local_testnet/README.md`
- The lost of `--debug-level` flag in `bootnode.sh`
- Avoid using GNU sed/grep in macOS system.

## Proposed Changes

Please list or describe the changes introduced by this PR.

## Additional Info

Please provide any additional information. For example, future considerations
or information useful for reviewers.

### CODE REVIEW

**Comment by michaelsproul** [comment]
which OS are you using where it didn't work? AFAIK our scripts are working on Linux and macOS. Are you using a BSD?

**Comment by mask-pp** [comment]
> which OS are you using where it didn't work? AFAIK our scripts are working on Linux and macOS. Are you using a BSD?

My OS is macOS m3.

**Comment by realbigsean** [comment]
> which OS are you using where it didn't work? AFAIK our scripts are working on Linux and macOS. Are you using a BSD?

On mac you need to install gnu sed and gnu grep to run these scripts. If these changes work for linux an mac out of the box, it'd be nice.

**Comment by realbigsean** [comment]
@mask-pp would you mind signing the CLA?

File: scripts/local_testnet/setup_time.sh [hunk 1]
```diff
@@ -20,16 +20,21 @@ echo "slot_per_epoch=$SLOT_PER_EPOCH"
 genesis_file=$1
 
 # Update future hardforks time in the EL genesis file based on the CL genesis time
-GENESIS_TIME=$(lcli pretty-ssz --spec $SPEC_PRESET --testnet-dir $TESTNET_DIR BeaconState $TESTNET_DIR/genesis.ssz | jq | grep -Po 'genesis_time": "\K.*\d')
-echo $GENESIS_TIME
+GENESIS_TIME=$(lcli pretty-ssz --spec $SPEC_PRESET --testnet-dir $TESTNET_DIR BeaconState $TESTNET_DIR/genesis.ssz | jq | sed -n 's/.*genesis_time": "\([0-9]*\).*/\1/p')
```
Comment (chong-he) [inline scripts/local_testnet/setup_time.sh:23]: The original script outputs the `genesis_time`, so does your script - they do the same thing. Not sure what do you mean by invalid option `--P`? Does the script not work in your OS (MacOS)?

File: scripts/local_testnet/bootnode.sh [hunk 2]
```diff
@@ -9,26 +9,13 @@ set -Eeuo pipefail
...
-echo "Generated bootnode enr and written to $TESTNET_DIR/boot_enr.yaml"
-
 DEBUG_LEVEL=${1:-info}
 
 echo "Starting bootnode"
 
-exec lighthouse boot_node \
+exec lighthouse \
+    --debug-level $DEBUG_LEVEL \
+    boot_node \
```
Comment (chong-he) [inline scripts/local_testnet/bootnode.sh:33]: Nice catch! The `DEBUG_LEVEL` wasn't used in the script previously, so it is nice to add it here

File: scripts/local_testnet/setup.sh [hunk 3]
```diff
@@ -51,3 +51,17 @@ lcli \
...
+	--ip 127.0.0.1 \
+	--udp-port $BOOTNODE_PORT \
+	--tcp-port $BOOTNODE_PORT \
+	--genesis-fork-version $GENESIS_FORK_VERSION \
+	--output-dir $DATADIR/bootnode
+
+bootnode_enr=`cat $DATADIR/bootnode/enr.dat`
+echo "- $bootnode_enr" > $TESTNET_DIR/boot_enr.yaml
+
+echo "Generated bootnode enr and written to $TESTNET_DIR/boot_enr.yaml"
```
Comment (chong-he) [inline scripts/local_testnet/setup.sh]: Not sure what is the motivation of moving the initialization of `boot_node` to `setup.sh`? There is no harm here, just that the original script will have:
`Now Generating bootnode enr` and
`Generated bootnode enr and written to $TESTNET_DIR` 
written to `bootnode.log`. Now it is written to `setup.log`

File: scripts/local_testnet/setup_time.sh [hunk 4]
```diff
@@ -20,16 +20,21 @@ echo "slot_per_epoch=$SLOT_PER_EPOCH"
...
+echo "GENESIS_TIME: $GENESIS_TIME"
+
 CAPELLA_TIME=$((GENESIS_TIME + (CAPELLA_FORK_EPOCH * $SLOT_PER_EPOCH * SECONDS_PER_SLOT)))
-echo $CAPELLA_TIME
-sed -i 's/"shanghaiTime".*$/"shanghaiTime": '"$CAPELLA_TIME"',/g' $genesis_file
+echo "SHANGHAI_TIME: $CAPELLA_TIME"
+sed 's/"shanghaiTime".*$/"shanghaiTime": '"$CAPELLA_TIME"',/g' $genesis_file > "/tmp/genesis.json"
```
Comment (chong-he) [inline scripts/local_testnet/setup_time.sh:28]: Does the command not work in MacOS (I am not familiar with it)? 

The original script with option `-i` in sed is to allow in-file editing straightaway, so it works. The proposed change creates a temporary file to replace the `genesis.json`, which does the same thing

**Review by chong-he — COMMENTED** [review]
Thanks for the PR, left some comments

**Comment by mask-pp** [comment]
> @mask-pp would you mind signing the CLA?

Done!

File: scripts/local_testnet/setup_time.sh (hunk 1 above)
Comment (mask-pp) [reply scripts/local_testnet/setup_time.sh:23]: > The original script outputs the `genesis_time`, so does your script - they do the same thing. Not sure what do you mean by invalid option `--P`? Does the script not work in your OS (MacOS)?

Yeah, it doesn't work in my MacOS.

File: scripts/local_testnet/setup_time.sh (hunk 4 above)
Comment (mask-pp) [reply scripts/local_testnet/setup_time.sh:28]: > Does the command not work in MacOS (I am not familiar with it)?
> 
> The original script with option `-i` in sed is to allow in-file editing straightaway, so it works. The proposed change creates a temporary file to replace the `genesis.json`, which does the same thing

It also doesn't work in MacOS, so I have to change it.

**Comment by mask-pp** [comment]
> Thanks for the PR, left some comments

Thanks very much for your review man.

**Comment by chong-he** [comment]
> > Thanks for the PR, left some comments
> 
> Thanks very much for your review man.

I am good with the proposed change since it does the same thing, and it will be good for other users using MacOS trying to run a local testnet using the scripts. So I am happy to recommend merging this PR. 

Thanks again!

Edit: how about the change in bootnode logs from to setup.sh: https://github.com/sigp/lighthouse/pull/5169#discussion_r1576198261

It is not a major thing, but may I know the motivation behind? I think that should work on MacOS? If so, we can remain that part the same

**Comment by mask-pp** [comment]
> > > Thanks for the PR, left some comments
> > 
> > 
> > Thanks very much for your review man.
> 
> I am good with the proposed change since it does the same thing, and it will be good for other users using MacOS trying to run a local testnet using the scripts. So I am happy to recommend merging this PR.
> 
> Thanks again!
> 
> Edit: how about the change in bootnode logs from to setup.sh: [#5169 (comment)](https://github.com/sigp/lighthouse/pull/5169#discussion_r1576198261)
> 
> It is not a major thing, but may I know the motivation behind? I think that should work on MacOS? If so, we can remain that part the same

Yeah, the `bootnode` change is optional. It is correct to write the bootnode log into bootnode.log. In order to reduce unnecessary changes I have reverted it.

**Comment by chong-he** [comment]
Thanks @mask-pp 

This seems to be a cleaner solution to enable MacOS to work with the script with `sed -i`: https://www.reddit.com/r/linux4noobs/comments/ne0ti8/why_does_sed_i_not_work_on_my_mac/

Could you give this a try and see if it works? If so, I can test linux on my end and we can use this cleaner solution

**Comment by mask-pp** [comment]
> Thanks @mask-pp
> 
> This seems to be a cleaner solution to enable MacOS to work with the script with `sed -i`: https://www.reddit.com/r/linux4noobs/comments/ne0ti8/why_does_sed_i_not_work_on_my_mac/
> 
> Could you give this a try and see if it works? If so, I can test linux on my end and we can use this cleaner solution

It works with `sed -i ''` in macOS, but it doesn't work in linux.
The best way I can imagine to be compatible with mac and linux is don't use the parameter "-i".

**Review by ackintosh — COMMENTED** [review]
Thanks for this PR!

I also noticed that we need to remove these lines from local-testnet.yml to avoid using GNU sed/grep in CI for macOS:
https://github.com/sigp/lighthouse/blob/8cd2b1ca87a7bf660130279558ed5dd9e91dda75/.github/workflows/local-testnet.yml#L42-L47

and the README in the local_testnet directory:
https://github.com/sigp/lighthouse/blob/003bb0ae3cc41e082c1e5462cfcd9341a4785bfc/scripts/local_testnet/README.md#L11

**Comment by mask-pp** [comment]
> Thanks for this PR!
> 
> I also noticed that we need to remove these lines from local-testnet.yml to avoid using GNU sed/grep in CI for macOS:
> 
> https://github.com/sigp/lighthouse/blob/8cd2b1ca87a7bf660130279558ed5dd9e91dda75/.github/workflows/local-testnet.yml#L42-L47
> 
> and the README in the local_testnet directory:
> 
> https://github.com/sigp/lighthouse/blob/003bb0ae3cc41e082c1e5462cfcd9341a4785bfc/scripts/local_testnet/README.md#L11

Deleted.

**Comment by chong-he** [comment]
I am not sure why we need to delete the line on the `README.md` file? @ackintosh 

As for the `local-testnet.yml`, does #5643 fix it?

**Comment by ackintosh** [comment]
> I am not sure why we need to delete the line on the README.md file? @ackintosh

If my understanding of this PR is correct, it eliminates the need for macOS users to install GNU sed/grep. They can now run start_local_testnet.sh using the built-in sed/grep commands. So the line in the README.md that encourages macOS users to install GNU sed/grep can be removed.

**Review by ackintosh — APPROVED** [review]
Thanks @mask-pp, LGTM.

Could you please update the `Proposed Changes` section of this [comment](https://github.com/sigp/lighthouse/pull/5169#issue-2112470323)?

**Comment by mask-pp** [comment]
> Thanks @mask-pp, LGTM.
> 
> Could you please update the `Proposed Changes` section of this [comment](https://github.com/sigp/lighthouse/pull/5169#issue-2112470323)?

changed.

**Comment by dapplion** [comment]
Thank you so much for your contribution @mask-pp, but we recently replaced our local testnet scripts with kurtosis
- https://github.com/sigp/lighthouse/pull/5865

**Pull request closed without merging**